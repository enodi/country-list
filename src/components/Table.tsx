import React from "react";
import numbro from "numbro";

import { COUNTRY_DETAIL, CURRENCY_DATA } from "../models/country.model";

type TableProps = {
    loading: boolean;
    countries: COUNTRY_DETAIL[];
};

const Table: React.FunctionComponent<TableProps> = ({ loading, countries }) => {
    const renderTable = () => {
        if (loading) {
            return (
                <tr>
                    <td>
                        <div className="loader__wrapper">
                            <span className="loading__anim"></span>
                        </div>
                    </td>
                </tr>
            );
        } else if (countries.length === 0) {
            return (
                <tr>
                    <td>
                        <div className="loader__wrapper">
                            <span>No Data</span>
                        </div>
                    </td>
                </tr>
            );
        } else {
            return countries.map((country: COUNTRY_DETAIL, index: number) => (
                <tr key={`country-${index}`}>
                    <td>{country.name}</td>
                    <td>{country.capital}</td>
                    <td>
                        {numbro(country.population).format({
                            thousandSeparated: true,
                        })}
                    </td>
                    {country.currencyData.map(
                        (currencies: CURRENCY_DATA, i: number) => {
                            if (currencies.currencyAmount) {
                                return (
                                    <React.Fragment
                                        key={`country-currency${i}`}
                                    >
                                        <td>{currencies.currency}</td>
                                        <td>
                                            <span className="symbol">
                                                {currencies.symbol}{" "}
                                            </span>
                                            {currencies.currencyAmount}
                                        </td>
                                    </React.Fragment>
                                );
                            }
                        }
                    )}
                    <td>
                        <img
                            src={country.flag}
                            alt="country flag"
                            className="dashboard--img-dimension"
                        />
                    </td>
                </tr>
            ));
        }
    };
    return (
        <table className="dashboard__table">
            <thead>
                <tr>
                    <th>Name</th>
                    <th>Capital</th>
                    <th>Population</th>
                    <th>Currency Code</th>
                    <th>Local Currency</th>
                    <th>Flag</th>
                </tr>
            </thead>
            <tbody>{renderTable()}</tbody>
        </table>
    );
};

export default Table;
