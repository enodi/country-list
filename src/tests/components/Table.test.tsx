import React from "react";
import { shallow } from "enzyme";

import Table from "../../components/Table";

const countryData = [
    {
        capital: "Abuja",
        currencyData: [{ currency: "NGN", currencyAmount: 45.424481 }],

        flag: "https://restcountries.eu/data/nga.svg",
        name: "Nigeria",
        population: 186988000,
    },
];

describe("Table Component", () => {
    let wrapper: any;
    beforeEach(() => {
        wrapper = shallow(<Table loading={false} countries={countryData} />);
    });

    it("should render Table component correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("should contain td element with text", () => {
        expect(wrapper.find("td").first().text()).toEqual("Nigeria");
    });
});
