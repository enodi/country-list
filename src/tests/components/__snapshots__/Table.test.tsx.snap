// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should render Table component correctly 1`] = `
<table
  className="dashboard__table"
>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Capital
      </th>
      <th>
        Population
      </th>
      <th>
        Currency Code
      </th>
      <th>
        Local Currency
      </th>
      <th>
        Flag
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="country-0"
    >
      <td>
        Nigeria
      </td>
      <td>
        Abuja
      </td>
      <td>
        186,988,000
      </td>
      <td>
        NGN
      </td>
      <td>
        45.424481
      </td>
      <td>
        <img
          alt="country flag"
          className="dashboard--img-dimension"
          src="https://restcountries.eu/data/nga.svg"
        />
      </td>
    </tr>
  </tbody>
</table>
`;
